<a href="https://colab.research.google.com/github/lostysky/Data-Engineering-codeboosters-internship-2026/blob/main/Day_03_ETL_Pandas_APIs.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install requests --quiet

In [ ]:
import pandas as pd
import numpy as np
import requests
import matplotlib.pyplot as plt


In [ ]:

import warnings
warnings.filterwarnings('ignore')

In [ ]:
print("Libraries imported successfully")
print(f'pandas version :{pd.__version__}')
print(f'request version :{requests.__version__}')

Libraries imported successfully
pandas version :2.2.2
request version :2.32.4


# Activity 1 -- Clean messy_sales_data.csv
Upload messy_sales_data.csv to Colab First panel, then run the cells below

Known data quality issue in this file :

1. Missing values in customer_name, quantity, category
2. Duplicates rows (orders 1001/1005 are identical)
3. Mixed data formats YYYY-MM-DD and DD-MM-YYYY
4. Inconsistent test case in customer_name (UPPER,LOWER,title)
5. Wrong category value (Keyborad labelled as Electronics)

In [ ]:
df=pd.read_csv('messy_sales_data.csv')
print("Dataset loaded successfully")

Dataset loaded successfully


In [ ]:
print("number of rows and column",df.shape)

number of rows and column (30, 9)


In [ ]:
df.columns.tolist()

['order_id',
 'customer_name',
 'product',
 'category',
 'quantity',
 'unit_price',
 'order_date',
 'city',
 'sales_rep']

In [ ]:

#=======================================================================================================
#CELL 3
#=======================================================================================================
print('='*55)
print('Data Quality Analysis')
print('='*55)
print(f'Number of rows : {df.shape[0]}')
print(f'Number of columns : {df.shape[1]}')


Data Quality Analysis
Number of rows : 30
Number of columns : 9


In [ ]:
# 1. Missing values
print('\n[1] MISSING VALUES PER COLUMN')
print(df.isnull().sum())
#.isnull() returns True/False for each cll
#.sum() count True (=missing) values per column


[1] MISSING VALUES PER COLUMN
order_id         0
customer_name    2
product          1
category         1
quantity         3
unit_price       0
order_date       0
city             0
sales_rep        0
dtype: int64


In [ ]:
# 2. Duplicates
print(f'\n[2] NUMBER OF DUPLICATES : {df.duplicated().sum()}')
#.duplicated() returns True for rows that are exact copies of a previous row
#.sum() counts how many duplicated rows exist


[2] NUMBER OF DUPLICATES : 0


In [ ]:
#3. Data types
print('\n[3] DATA TYPES :')
print(df.dtypes)


[3] DATA TYPES :
order_id           int64
customer_name     object
product           object
category          object
quantity         float64
unit_price         int64
order_date        object
city              object
sales_rep         object
dtype: object


In [ ]:

#4. Unique values in text columns (sopt inconsistencies)
print('\n[4] UNIQUE CATEGORIES:',df['category'].unique())
print('\n[4] sample customer names :',df['customer_name'].dropna().unique()[:8])
print('\n[4] sample order_date values:',df['order_date'].unique()[:6])


[4] UNIQUE CATEGORIES: ['Electronics' 'Accessories' nan]

[4] sample customer names : ['Ramesh Kumar' 'Priya Nair' 'AMIT VERMA' 'Sunita Patel' 'kiran mehta'
 'Deepak Singh' 'Ananya Das' 'Vikram Iyer']

[4] sample order_date values: ['2024-01-05' '2024-01-07' '2024-01-08' '2024-01-10' '07-01-2024'
 '2024-01-12']


In [ ]:
#=======================================================================================================
#CELL 4 - Create a working copy(ETL best practice)
#=======================================================================================================
copy=df.copy()
#.copy() creates a completely independent copy of the dataframe
copy
#This is the most important ETL hygiene rule

,order_id,customer_name,product,category,quantity,unit_price,order_date,city,sales_rep
0,1001,Ramesh Kumar,Laptop,Electronics,2.0,45000,2024-01-05,Mumbai,Anil Sharma
1,1002,Priya Nair,NaN,Electronics,1.0,15000,2024-01-07,Delhi,Sunita Rao
2,1003,AMIT VERMA,Keyboard,Accessories,3.0,1200,2024-01-08,Bangalore,Anil Sharma
3,1004,Sunita Patel,Monitor,Electronics,NaN,22000,2024-01-10,Chennai,Ravi Kumar
4,1005,Ramesh Kumar,Laptop,Electronics,2.0,45000,2024-01-05,Mumbai,Anil Sharma
5,1006,kiran mehta,Mouse,Accessories,10.0,800,07-01-2024,Pune,Sunita Rao
6,1007,Deepak Singh,Headphones,Electronics,2.0,3500,2024-01-12,Hyderabad,Ravi Kumar
7,1008,NaN,Webcam,Accessories,1.0,2500,2024-01-13,Mumbai,Anil Sharma
8,1009,Ananya Das,Laptop,Electronics,1.0,45000,2024-01-15,Kolkata,Sunita Rao
9,1010,Vikram Iyer,Keyboard,Accessories,5.0,1200,2024-01-15,Chennai,Ravi Kumar


In [ ]:
df['customer_name'].fillna('Unknown Customer',inplace=True)

In [ ]:
df['customer_name'].tolist()


['Ramesh Kumar',
 'Priya Nair',
 'AMIT VERMA',
 'Sunita Patel',
 'Ramesh Kumar',
 'kiran mehta',
 'Deepak Singh',
 'Unknown Customer',
 'Ananya Das',
 'Vikram Iyer',
 'Pooja Gupta',
 'SURESH RAO',
 'Meera Joshi',
 'Arjun Nair',
 'Tanvi Mehta',
 'Kiran Mehta',
 'Deepak Singh',
 'Rohit Verma',
 'Sneha Reddy',
 'Gaurav Shukla',
 'Nisha Kapoor',
 'Ajay Tiwari',
 'ANANYA DAS',
 'Preeti Saxena',
 'Amit Bose',
 'Unknown Customer',
 'Rekha Nair',
 'Harish Pillai',
 'Sanjay Dubey',
 'Kavya Nambiar']

In [ ]:
median_qty=df['quantity'].median()

In [ ]:

df['quantity'].isnull().sum()

np.int64(3)

In [ ]:
df['category'].fillna('uncategorized',inplace=True)
df['quantity'].tolist()


[2.0,
 1.0,
 3.0,
 nan,
 2.0,
 10.0,
 2.0,
 1.0,
 1.0,
 5.0,
 2.0,
 8.0,
 nan,
 3.0,
 2.0,
 10.0,
 5.0,
 1.0,
 1.0,
 15.0,
 2.0,
 3.0,
 1.0,
 4.0,
 6.0,
 1.0,
 2.0,
 3.0,
 10.0,
 nan]

In [ ]:

df['category'].isnull().sum()


np.int64(0)

In [ ]:

df['category'].tolist()

['Electronics',
 'Electronics',
 'Accessories',
 'Electronics',
 'Electronics',
 'Accessories',
 'Electronics',
 'Accessories',
 'Electronics',
 'Accessories',
 'Electronics',
 'Accessories',
 'Electronics',
 'Electronics',
 'Accessories',
 'Accessories',
 'Accessories',
 'Electronics',
 'Electronics',
 'Accessories',
 'Electronics',
 'Accessories',
 'Electronics',
 'Electronics',
 'Accessories',
 'Electronics',
 'uncategorized',
 'Electronics',
 'Accessories',
 'Accessories']

In [ ]:
df['category'].fillna("UnCategorized",inplace=True)
df['category']

,category
0,Electronics
1,Electronics
2,Accessories
3,Electronics
4,Electronics
5,Accessories
6,Electronics
7,Accessories
8,Electronics
9,Accessories


In [ ]:
df.isnull().sum().sum()

np.int64(4)

In [ ]:

df['product'].fillna("Unknown Product",inplace=True)

In [ ]:
#CELL 6
print(f'Before deduplication :{len(df)} rows')
print(f'Duplicate rows : {df.duplicated().sum()}')

#Shows which rows are duplicates
print('\nDuplicate rows')
print(df[df.duplicated(keep=False)][['order_id','customer_name','product','order_date']])

#keep=False marks ALL copies of a duplicate (not just the seconds)
#this lets us SEE both the original and the duplicates before removing

#remove exact duplicates rows
df.drop_duplicates(inplace=True)
#drop_duplicates() removes rows that are IDENTICAL in ALL columns
#This lets us SEE both the original and the duplicate before removing

print(f'After deduplication :{len(df)}')

Before deduplication :30 rows
Duplicate rows : 0

Duplicate rows
Empty DataFrame
Columns: [order_id, customer_name, product, order_date]
Index: []
After deduplication :30


In [ ]:
print(f'After deduplication :{len(copy)} rows')
print(f'Rows Removed: {len(df)-len(copy)}')


After deduplication :30 rows
Rows Removed: 0


In [ ]:
#CELL 6

print('Sample dates before parsing:')
print(df['order_date'].head(8).tolist())
#Some are YYYY-MM-DD and some are DD-MM-YYYY

df['order_date']=pd.to_datetime(
    df['order_date'],
    dayfirst=False,    #Try YYYY-MM-DD format first
    errors='coerce'    #If parsing fails, put NaT (Not a Time) instead of crashing
    )

#pd.to_datetime() converts strings to proper datetime objects
#dayfirst=False: treats YYYY-MM-DD as a primary format
#errors='coerce': crictical  -- without this  one bad date crashes the entire column

Sample dates before parsing:
['2024-01-05', '2024-01-07', '2024-01-08', '2024-01-10', '2024-01-05', '07-01-2024', '2024-01-12', '2024-01-13']


In [ ]:
#Check for any unparseable dates (would show as NAT)
nat_count=df['order_date'].isnull().sum()
print(f'\nUnparseable dates (NaT): {nat_count}')

#Extract date components - very useful for analysis
df['year']=df['order_date'].dt.year
df['month']=df['order_date'].dt.month
df['month_name']=df['order_date'].dt.strftime('%B')
#.dt is the datetime accessor - gives access to year, month, day,etc
#.strtime('%B) returns the full month name:'Janunary','February',etc.

print('\nSample dates after parsing:')
print(df[['order_date','year','month','month_name']])



Unparseable dates (NaT): 2

Sample dates after parsing:
   order_date    year  month month_name
0  2024-01-05  2024.0    1.0    January
1  2024-01-07  2024.0    1.0    January
2  2024-01-08  2024.0    1.0    January
3  2024-01-10  2024.0    1.0    January
4  2024-01-05  2024.0    1.0    January
5         NaT     NaN    NaN        NaN
6  2024-01-12  2024.0    1.0    January
7  2024-01-13  2024.0    1.0    January
8  2024-01-15  2024.0    1.0    January
9  2024-01-15  2024.0    1.0    January
10 2024-01-18  2024.0    1.0    January
11 2024-01-20  2024.0    1.0    January
12 2024-01-22  2024.0    1.0    January
13        NaT     NaN    NaN        NaN
14 2024-01-25  2024.0    1.0    January
15 2024-01-07  2024.0    1.0    January
16 2024-01-28  2024.0    1.0    January
17 2024-01-30  2024.0    1.0    January
18 2024-02-01  2024.0    2.0   February
19 2024-02-03  2024.0    2.0   February
20 2024-02-05  2024.0    2.0   February
21 2024-02-07  2024.0    2.0   February
22 2024-01-15  2024.0  

In [ ]:
#Fix 4 standardize customer names to title case
print('Before Standardization:',df['customer_name'].unique()[:6])

df['customer_name']=(
    df['customer_name']
    .str.strip()     #remove leading/trailing whitespace
    .str.title()     #convert to title case:'amit verma' ->'Amit Verma'
)

#.str is pandas string accessor - applies string methods to every row
#.strip() removes spaces at edges: '   Ramesh ' ->'ramesh'
#.title() capitalizes first letter of each word
#Without this, 'AMIT VERMA' and 'Amit Verma' look like different customers

print('\nAfter Standardization:',df['customer_name'].unique()[:6])
#

Before Standardization: ['Ramesh Kumar' 'Priya Nair' 'AMIT VERMA' 'Sunita Patel' 'kiran mehta'
 'Deepak Singh']

After Standardization: ['Ramesh Kumar' 'Priya Nair' 'Amit Verma' 'Sunita Patel' 'Kiran Mehta'
 'Deepak Singh']


In [ ]:
#Fis 5: Correct wrong category for keyboard rows
print(f'\nBefore: Keyboard rows with Electronics Category:')
wrong_mask=(df['product']=='Keyboard') & (df['category']=='Electronics')
#wrong_mask is a boolean series: true where BOTH conditions are met
#& is the bitwise AND operator for combining boolean conditions in Pandas
print(df[wrong_mask][['product','category']])

df.loc[wrong_mask,'category']='Accessories'
#df.loc[row_mask,column_name]=value
#This sets 'category' to 'Accessories' only for rows where wrong_mask is True
#All other rows are unaffected

print('After fix: Unique categories:',df['category'].unique())


Before: Keyboard rows with Electronics Category:
     product     category
23  Keyboard  Electronics
After fix: Unique categories: ['Electronics' 'Accessories' 'uncategorized']


In [ ]:

#CELL 9
#Ensure numeric columns have the right types
df['quantity']=pd.to_numeric(df['quantity'],errors='coerce').fillna(median_qty).astype(int)
df['unit_price']=pd.to_numeric(df['unit_price'],errors='coerce')
#pd.to_numeric() converts any column to a numeric type
#error='coerce' turns unconvertible values to NaN instead of crashing
#.astype(int) converts float(2.0) to integer (2) for quantity

#Create revenue column:quantity * unit_price
df['revenue']=df['quantity']*df['unit_price']

#This is a vectorized iperation:quantity - pandas multiplies row by row automatically
#just like df['total_score'] = df['math']+df['Science'] from Day 1

print('Revenue column created')
print(df[['customer_name','product','quantity','unit_price','revenue']].head(5))
print(f'\nTotal Revenue across all orders: Rs.{df["revenue"].sum():,.0f}')
df[['customer_name','product','quantity','unit_price','revenue']]

Revenue column created
  customer_name          product  quantity  unit_price  revenue
0  Ramesh Kumar           Laptop         2       45000    90000
1    Priya Nair  Unknown Product         1       15000    15000
2    Amit Verma         Keyboard         3        1200     3600
3  Sunita Patel          Monitor         2       22000    44000
4  Ramesh Kumar           Laptop         2       45000    90000

Total Revenue across all orders: Rs.818,000


,customer_name,product,quantity,unit_price,revenue
0,Ramesh Kumar,Laptop,2,45000,90000
1,Priya Nair,Unknown Product,1,15000,15000
2,Amit Verma,Keyboard,3,1200,3600
3,Sunita Patel,Monitor,2,22000,44000
4,Ramesh Kumar,Laptop,2,45000,90000
5,Kiran Mehta,Mouse,10,800,8000
6,Deepak Singh,Headphones,2,3500,7000
7,Unknown Customer,Webcam,1,2500,2500
8,Ananya Das,Laptop,1,45000,45000
9,Vikram Iyer,Keyboard,5,1200,6000


In [ ]:
#CELL 10  - Validate Final data quality check
print('='*55)
print('     POST CLEANING VALIDATION REPORT     ')
print('='*55)
print(f'Original rows   : {len(copy)}')
print(f'Cleaned rows    : {len(df)}')
print(f'Rows Removed    : {len(copy)-len(df)}(duplicates)')
print(f'Missing values  : {df.isnull().sum().sum()}')
print(f'Duplicates      : {df.duplicated().sum()}')
print(f'Date nulls      : {df['order_date'].isnull().sum()}')
print(f'Revenue NaN     : {df['revenue'].isnull().sum()}')
print(f'Categories      : {sorted(df["category"].unique())}')
print('='*55)

#ALL checks should show clean results
all_clean=(
    df.isnull().sum().sum()==0 and
    df.duplicated().sum()==0
)
print(f'DATA IS CLEAN : {all_clean}')

     POST CLEANING VALIDATION REPORT     
Original rows   : 30
Cleaned rows    : 30
Rows Removed    : 0(duplicates)
Missing values  : 8
Duplicates      : 0
Date nulls      : 2
Revenue NaN     : 0
Categories      : ['Accessories', 'Electronics', 'uncategorized']
DATA IS CLEAN : False


In [ ]:

print('Missing values before final cleaning:')
print(df.isnull().sum())

Missing values before final cleaning:
order_id         0
customer_name    0
product          0
category         0
quantity         0
unit_price       0
order_date       2
city             0
sales_rep        0
year             2
month            2
month_name       2
revenue          0
dtype: int64


In [ ]:
# Fill NaT values in 'order_date' with the mode of the column
mode_order_date = df['order_date'].mode()[0]
df['order_date'].fillna(mode_order_date, inplace=True)

# Re-extract date components after filling NaT
df['year'] = df['order_date'].dt.year
df['month'] = df['order_date'].dt.month
df['month_name'] = df['order_date'].dt.strftime('%B')

# Fill the single NaN in 'product' column with 'Unknown Product'
df['product'].fillna('Unknown Product', inplace=True)

print('\nMissing values after final cleaning:')
print(df.isnull().sum())


Missing values after final cleaning:
order_id         0
customer_name    0
product          0
category         0
quantity         0
unit_price       0
order_date       0
city             0
sales_rep        0
year             0
month            0
month_name       0
revenue          0
dtype: int64


In [ ]:
# CELL 10  - Validate Final data quality check (re-run)
print('='*55)
print('     POST CLEANING VALIDATION REPORT     ')
print('='*55)
print(f'Original rows   : {len(copy)}')
print(f'Cleaned rows    : {len(df)}')
print(f'Rows Removed    : {len(copy)-len(df)}(duplicates)')
print(f'Missing values  : {df.isnull().sum().sum()}')
print(f'Duplicates      : {df.duplicated().sum()}')
print(f'Date nulls      : {df['order_date'].isnull().sum()}')
print(f'Revenue NaN     : {df['revenue'].isnull().sum()}')
print(f'Categories      : {sorted(df["category"].unique())}')
print('='*55)

#ALL checks should show clean results
all_clean=(
    df.isnull().sum().sum()==0 and
    df.duplicated().sum()==0
)
print(f'DATA IS CLEAN : {all_clean}')

     POST CLEANING VALIDATION REPORT     
Original rows   : 30
Cleaned rows    : 30
Rows Removed    : 0(duplicates)
Missing values  : 0
Duplicates      : 0
Date nulls      : 0
Revenue NaN     : 0
Categories      : ['Accessories', 'Electronics', 'uncategorized']
DATA IS CLEAN : True


In [ ]:
#Revenue by product
product_rev=(
    df.groupby('product')
    ['revenue']
    .sum()
    .sort_values(ascending=False)
)

#.groupby('product)  - group all rows by product name
#['revenue']         - select the revenue column from each group
# .sum()             - add up revenue for each group
#
print('Revenue by product')
print(product_rev.to_string(index=False))

#Multi-metric aggregation using agg()
category_summary=df.groupby('category').agg(
    total_revenue  =('revenue','sum'),
    avg_order_value=('revenue','mean'),
    num_orders=('order_id','count'),
    unique_products=('product','nunique')
).round(2).reset_index()
#agg() applies MULTIPLE functions to MULTIPLE columns in one cell
#Format: new_col_name=('source_column','aggregation_function')
#'nunique' = count of unique values (not built into groupby directly)

print('\nCategory Summary')
print(category_summary)

Revenue by product
product
540000
154000
 28000
 20800
 20400
 20000
 19800
 15000

Category Summary
        category  total_revenue  avg_order_value  num_orders  unique_products
0    Accessories          81000          5785.71          14                4
1    Electronics         693000         46200.00          15                4
2  uncategorized          44000         44000.00           1                1


In [ ]:
#CELL 12
df.to_csv('clean_sales_data.csv',index=False)
#to_csv() saves the DataFrame to a CSV file
#'clean_sales_data.csv' - the output filename
#index=False  - do NOT write the row numbers (0,1,2...) as a column
#This is the LOAD step of our ETL pipeline

print('CLeaned data saved to: clean_sales_data.csv')
print(f'Final dataset: {df.shape[0]} rows * {df.shape[1]} columns')
print('\nETL Pipeline for Sales Data: COMPLETE')
print(' EXTRACT -> messy_sales_data.csv loaded')
print(' TRANSFORM -> messy_sales_data.csv cleaned')
print(' LOAD -> clean_sales_data.csv saved')


CLeaned data saved to: clean_sales_data.csv
Final dataset: 30 rows * 13 columns

ETL Pipeline for Sales Data: COMPLETE
 EXTRACT -> messy_sales_data.csv loaded
 TRANSFORM -> messy_sales_data.csv cleaned
 LOAD -> clean_sales_data.csv saved


In [ ]:
#CELL 13
API_KEY='d2c9dfb04be6d96cd613967a87e684a0'
#Replace you API_KEY_HERE with the actual openweathermap API key
#A free key allows 60 calls per minute - more than enough for this activity

BASE_URL='https://api.openweathermap.org/data/2.5/weather'
#This is the API endpoint URL
#'data/2.5/weather'= the specific API route for currebt weather

#Cities to query - 8 major indian cities
CITIES=['Mumbai','Delhi','Bangalore','Chennai','Hyderabad','Ahmedabad','Kolkata','Pune','Jaipur']
#we will call the API once per city - 8 API calls total

print(f'API configured for {len(CITIES)} cities')
import requests
import pandas as pd

# --- EXTRACT --- #

API_KEY = 'c9dfb04be6d96cd613967a87e684a0' # Replace with your actual OpenWeatherMap API key
BASE_URL = 'http://api.openweathermap.org/data/2.5/weather'
CITIES = ['Mumbai', 'Delhi', 'Bangalore', 'Chennai', 'Hyderabad', 'Kolkata', 'Pune', 'Jaipur']

def fetch_weather(city, API_KEY):
    """
    Fetch current weather data for a given city.
    Returns a dictionary with weather metrics, or None on failure.
    """
    params = {
        'q':     city,
        'appid': API_KEY,
        'units': 'metric'
    }
    try:
        response = requests.get(BASE_URL, params=params, timeout=10)
        if response.status_code == 200:
            data = response.json()
            return {
                'city':        city,
                'temperature': round(data['main']['temp'], 1),
                'feels_like':  round(data['main']['feels_like'], 1),
                'humidity':    data['main']['humidity'],
                'pressure':    data['main']['pressure'],
                'wind_speed':  data['wind']['speed'],
                'condition':   data['weather'][0]['description'].title(),
                'visibility':  data.get('visibility', 0) // 1000
            }
        else:
            print(f'  ERROR {response.status_code} for {city}: {response.json().get("message","unknown error")}')
            return None

    except requests.exceptions.ConnectionError:
        print(f'  CONNECTION ERROR for {city} — check internet connection')
        return None
    except requests.exceptions.Timeout:
        print(f'  TIMEOUT for {city} — API did not respond in 10 seconds')
        return None


print('Calling Weather API...')
weather_records = []
for city in CITIES:
    print(f'  Fetching: {city}...', end='')
    record = fetch_weather(city, API_KEY)
    if record:
        weather_records.append(record)
        print(f' {record["temperature"]}°C, {record["condition"]}')
    else:
        print(' FAILED')

print(f'\nSuccessfully fetched: {len(weather_records)}/{len(CITIES)} cities')

# --- TRANSFORM --- #
# Create DataFrame from API results
weather_df = pd.DataFrame(weather_records)

print('\nWeather DataFrame created and data cleaned (structured):')
print(weather_df.head())
print(f'Shape: {weather_df.shape}')
print(f'Missing values: {weather_df.isnull().sum().sum()}')

# --- LOAD --- #

# Save the DataFrame to a CSV file
output_filename = 'weather_data_etl_output.csv'
weather_df.to_csv(output_filename, index=False)

print(f'\nWeather data saved to: {output_filename}')
print('\nWeather Data ETL Pipeline: COMPLETE')
print('  EXTRACT   → OpenWeatherMap API called')
print('  TRANSFORM → JSON parsed, DataFrame built, units converted')
print(f'  LOAD      → {output_filename} saved')


API configured for 9 cities
Calling Weather API...
  Fetching: Mumbai...  ERROR 401 for Mumbai: Invalid API key. Please see https://openweathermap.org/faq#error401 for more info.
 FAILED
  Fetching: Delhi...  ERROR 401 for Delhi: Invalid API key. Please see https://openweathermap.org/faq#error401 for more info.
 FAILED
  Fetching: Bangalore...  ERROR 401 for Bangalore: Invalid API key. Please see https://openweathermap.org/faq#error401 for more info.
 FAILED
  Fetching: Chennai...  ERROR 401 for Chennai: Invalid API key. Please see https://openweathermap.org/faq#error401 for more info.
 FAILED
  Fetching: Hyderabad...  ERROR 401 for Hyderabad: Invalid API key. Please see https://openweathermap.org/faq#error401 for more info.
 FAILED
  Fetching: Kolkata...  ERROR 401 for Kolkata: Invalid API key. Please see https://openweathermap.org/faq#error401 for more info.
 FAILED
  Fetching: Pune...  ERROR 401 for Pune: Invalid API key. Please see https://openweathermap.org/faq#error401 for more i

# Practice question
Q1. What are the three stages of ETL? Describe each stage using an example from today's sales dataset

Q2. A DataFrame has 500 rows. After calling df.dropna() it has 42 rows what does this tell you?

Q3. Write code to remove deplicates from df where same row means same customer_name and same product

Q4. What is the difference between fillna(0) and fillna(df['col'].median())? when would you perfer each?

Q5. Write Python code to call the weather API for 'Delhi' and print the temperature in Celsius

Q6.What does response.status.code==200 mean? what should you do when the code is 401

In [ ]:
  import pandas as pd

# --- CONFIGURATION --- #
excel_file_path = 'employee_salary_data.xlsx' # Make sure this Excel file is uploaded or accessible
output_csv_file = 'cleaned_employee_salary.csv'

# --- EXTRACT --- #
print('--- Starting Employee Salary ETL Project ---')

try:
    # Attempt to load data from Excel
    df_raw = pd.read_excel(excel_file_path)
    print(f'Successfully loaded data from {excel_file_path}. Shape: {df_raw.shape}')
except FileNotFoundError:
    print(f'Warning: {excel_file_path} not found. Creating dummy data for demonstration.')
    # Create a dummy DataFrame if the Excel file is not found
    data = {
        'Employee ID': [101, 102, 103, 104, 105, 101, 106, 107, 108],
        'Name': ['Alice Smith', 'Bob Johnson', 'Charlie Brown', 'Diana Prince', 'Eve Adams', 'Alice Smith', 'Frank White', 'Grace Black', 'Harry Green'],
        'Department': ['HR', 'IT', 'Finance', 'Marketing', 'HR', 'HR', 'IT', 'Finance', 'Marketing'],
        'Monthly Salary': [5000, 6000, 7500, 5500, 5200, 5000, 6200, 7800, 5800]
    }
    df_raw = pd.DataFrame(data)
    print(f'Dummy DataFrame created. Shape: {df_raw.shape}')

df = df_raw.copy() # Create a working copy

# --- TRANSFORM --- #
print('\n--- Data Transformation ---')

# 1. Remove duplicate records
print(f'Before dropping duplicates: {len(df)} rows')
duplicates_before = df.duplicated().sum()
df.drop_duplicates(inplace=True)
print(f'Removed {duplicates_before} duplicate rows. After dropping duplicates: {len(df)} rows')

# 2. Calculate Yearly Salary
if 'Monthly Salary' in df.columns:
    df['Yearly Salary'] = df['Monthly Salary'] * 12
    print('Calculated "Yearly Salary" based on "Monthly Salary".')
elif 'salary' in df.columns:
    df['Yearly Salary'] = df['salary'] * 12
    print('Calculated "Yearly Salary" based on "salary".')
else:
    print('Could not find "Monthly Salary" or "salary" column to calculate yearly salary.')

print('\nTransformed Data (first 5 rows):')
print(df.head())
print(f'Final DataFrame shape after transformations: {df.shape}')
print(f'Missing values after transformations: {df.isnull().sum().sum()}')

# --- LOAD --- #
print('\n--- Data Loading ---')
df.to_csv(output_csv_file, index=False)
print(f'Cleaned employee salary data saved to: {output_csv_file}')

print('\n--- Employee Salary ETL Project: COMPLETE ---')
print('  EXTRACT   → Data loaded from Excel (or dummy data generated)')
print('  TRANSFORM → Duplicates removed, Yearly Salary calculated')
print(f'  LOAD      → {output_csv_file} saved')

--- Starting Employee Salary ETL Project ---
Dummy DataFrame created. Shape: (9, 4)

--- Data Transformation ---
Before dropping duplicates: 9 rows
Removed 1 duplicate rows. After dropping duplicates: 8 rows
Calculated "Yearly Salary" based on "Monthly Salary".

Transformed Data (first 5 rows):
   Employee ID           Name Department  Monthly Salary  Yearly Salary
0          101    Alice Smith         HR            5000          60000
1          102    Bob Johnson         IT            6000          72000
2          103  Charlie Brown    Finance            7500          90000
3          104   Diana Prince  Marketing            5500          66000
4          105      Eve Adams         HR            5200          62400
Final DataFrame shape after transformations: (8, 5)
Missing values after transformations: 0

--- Data Loading ---
Cleaned employee salary data saved to: cleaned_employee_salary.csv

--- Employee Salary ETL Project: COMPLETE ---
  EXTRACT   → Data loaded from Excel (or dum